chanfitness/chanfitness.go

```diff
@@ -0,0 +1,300 @@
+// Package chanfitness monitors the behaviour of channels to provide scores
+// which can be factored into the decision to close a channel. This is
+// achieved by maintaining an event store which tracks events for each
+// channel.
+//
+// Lifetime: the total time the channel has been known to the scoring system.
+// Note that this does not equal channel lifetime because data is not currently
+// persisted.
+//
+// Uptime: the total time within the channel's lifetime that its peer has been
+// observed as being online.
+package chanfitness
+
+import (
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/channelnotifier"
+	"github.com/lightningnetwork/lnd/peernotifier"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// ChannelEventStore maintains a set of event logs for the node's channels.
+// It is intended to provide insight into the performance of channels.
+type ChannelEventStore struct {
+	cfg *Config
+
+	// channels maps short channel IDs to event logs.
+	channels map[uint64]*chanEventLog
+
+	quit chan struct{}
+
+	sync.Mutex
+
+	wg sync.WaitGroup
+}
+
+// Config provides the event store with functions required to monitor channel
+// activity. All elements of the config must be non-nil for the event store to
+// operate.
+type Config struct {
+	// Channel events provides a subscription client which provides a
+	// stream of channel events.
+	ChannelEvents func() (*subscribe.Client, error)
+
+	// PeerEvents provides a subscription client which provides a stream of
+	// peer online/offline events.
+	PeerEvents func() (*subscribe.Client, error)
```

Comment:
If these are untyped anyway, and especially if this is extended to a large list in the future, can't the store not just have a single channel over which it receives all events? And then have all event sources drop their typed events into that channel.

Reply:
So we'd construct a function which pipes all of the subscriptions into a channel which we create in `server.go` and then pass that to `Config`? I think that constructing that kind of aggregation from the caller bubbles complexity up rather than down sometimes, because if we have a lot of event sources it's simpler to just provide a list of functions than have some long aggregations of them all in the middle of the `newServer` call? 

Having had my philosophical ramble, I think this may address some of the concerns you have about the unit tests so will give this a try :)

Reply:
I was actually thinking something simpler. Just make the fitness system a dependencies of the other subsystems. It exposes a single function `SendEvent` to those subsystems. In the implementation of `SendEvent`, the only thing that happens is that the (untyped) event is dropped into a single buffered channel or queue. The main loop of the fitness system is receiving on that channel and processing the events via a type switch.

But I could be missing something here of course.

Reply:
My main concern with a `SendEvent` call is that it feels a bit counterintuitive to be sending events when we have two subscriptions specifically intended for receiving events. 

In the case where we don't have subscriptions, a `SendEvent` makes sense. And I suspect this will be the case for the remainder of the events we add to channel scores (ie htlc level events). 

Since it probably (?) doesn't make sense to make subscriptions for htlc level events the options would be:
1. Use `SendEvent` everywhere, accepting that it's a bit of an off construction to send events in the cases where you have subscriptions available
2. Use `SendEvent` for non-subscription events and still pass the Peer/Chan Subscriptions in, but pipe those into the SendEvent channel so that monitoring happens in one central place.

Reply:
Isn't it that these subscriptions only exist to serve the event store? In `server.go`, the call `s.peerNotifier.NotifyPeerOnline(pubKey)` is made, which could be replaced by a call directly into the event store `SendEvent`?

I am looking for the simplest thing that works, but may miss something because I didn't follow the design decision around peer notifier.

Reply:
`PeerNotifier` does exist to serve the event store, so that could be replaced. but `ChannelNotifier` is more general. I thought that it would be better to have a central subscription queuing up events, rather than having calls out to `chanFitness` in the middle of other functions, but it would really only be out of place in `channelNotifier.NotifyOpen/ClosedChannelEvent` which I guess is ok.

Reply:
In the implementation of that function `chanFitness.SendEvent`, the only thing that would happen is queue the event in a non-blocking way. 

I don't want to give the impression that I know what the best design is. Just wanted to bring up this alternative, because I was wondering how the current approach would scale. If for many types of events a separate connecting component is required that basically just inverts the control flow, it may create too much boilerplate code compared to sending the event to `chanFitness` directly.

Reply:
Of course, I'm glad you brought it up because I hadn't really thought what this will look like for events which don't have subscriptions available. You're right about not wanting to add large amounts of code just to subscribe to different kinds of events events. And unit testing with subscriptions has been tricky (which often means the design isn't quite right, I find) so I'm going to try with a SendEvent structure and see how that looks :)

Reply:
The amount of code needed to subscribe an event is pretty minor. We created the entire `subscribe` package in the first place in order to consolidate all the pub/sub (and infinite queue) duplication we had strewn around the codebase (chain notifier, etc). It also takes care of a number of details around client tracking, event cancellation, etc. All that would need to be replicated again vs just using an existing package that is meant to serve as a building block for sub-systems like this one. The model is essentially one of a messaging hub where any sub-system can listen on an even stream with minimal overhead.

Reply:
So the current design (taking peer notification as an example) is:
* `server` calls `peernotifier.NotifyPeerOnline`
* `peernotifier` pushes update into a channel to `subscribe.Server`
* `subscribe.Server` pushes update into a channel to `chanfitness`

What is the advantage of this over:
* `server` calls `chanfitness.NotifyPeerOnline`
* `chanfitness` internally manages a queue to not block reporting sub-systems. There only needs to be a single queue for all events.

I only see an advantage of having separate `*-notifier` objects if there are multiple listeners. Those can be added when a need arises. From the sub-system perspective the only required change would then be to call the `notifier` instead of `chanfitness`, which is just a replacement without any other design implications for that sub-system.

I think we should be careful not to over-engineer this.

Reply:
In my mind there are three sources that the scoring system will need events from channels, peers and HTLCs. And justifying using a subscribe server requires that we have more than one listener per subscription, otherwise we're just writing boilerplate as you say. 

1. `ChanNotifier`: already used elsewhere, so adding `SendEvent` call would be a duplication
2. `PeerNotifier`: created for this PR, only other use at present is potentially a rpc subscription
3. HTLC events: no subscription at present, although @Roasbeef has recently opened #3420

I am reluctant to start off with an event queue if it is inevitably going to be replaced by subscriptions, so I think the questions that need answering are:
1. Are there use cases for `PeerNotifier` and a potential `HTLCNotifier` and is a user exposed RPC call a good enough justification for the existence of a subscription? 
2. How likely is it that we extend the scoring system beyond these three sources? Because if there's anything else would reasonably want to add that wouldn't suit a subscription that's an argument for an events queue 

I'm not familiar enough with the repo to make a call about whether subscriptions for these things are required. I would like to move forward with this PR, so if I can get some clarity on those questions I think it will help determine what direction to move in?

Reply:
Although an events chan is looking a lot cleaner for tests. The only ugliness is that it has to be called from the channel open/close notification functions which feels like doing the same thing twice. So it really is dependent on whether subscriptions actually have other use cases. which I'm unsure of.

Reply:
> chanfitness internally manages a queue to not block reporting sub-systems. There only needs to be a single queue for all events.


It would more or less need to re-implement the current queue handling in the `subscribe` package to do this. Therefore we should re-use what we have rather than attempt to over optimize for this use case which isn't performance critical and still nascent.

Reply:
I don't think we need to get caught up in this relatively minor difference. The original primary reviewer (johan) was OK with the current approach. We shouldn't undo that prior lineage due to a new reviewer stepping in. For the sake of making progress, I don't think there's a fundamental principle that tends us in one direction or the other. Therefore we can continue down the current path which had already received review by the former primary reviewer. 

If this was more critical code such as the channel state machine or revocation handling, then I would be willing to spend more time here, but it's a relatively independent sub-system that is non-critical.

Reply:
Even though Johan was ok with the current approach, it could still be that he likes the alternative more. I don't think we've discussed that. But yes, both approaches lead to functioning, correct code. This is only about preventing boiler plate. I can put aside my concerns about that and see how it works out.

Reply:
I think the current approach is more general and would scale better with more subscription types, and more subscribers to the different types. 

We had to take a similar decision when we initially added the `NotifyPeerOnline` API to the `fundingmanager`. We could have made the server call into the funding manager directly, but rather we made it a general subscription API that any subsystem could call. This has turned out useful to have access to also in other packages.

I also don't see the boiler plate we would avoid, since it is mostly contained in the `subscribe`  package?